Propagate errors for saveSettings action

jorgemd24 opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

This is a follow-up of https://github.com/woocommerce/google-listings-and-ads/issues/1331#issuecomment-1109566842 , https://github.com/woocommerce/google-listings-and-ads/issues/1331#issuecomment-1110084066 & https://github.com/woocommerce/google-listings-and-ads/pull/1586

This PR removes the try and catch from `saveSettings` action, allowing the error to bubble up and be handled by the consumer. if `saveSettings` were failing, the consumer was not able to tell if `saveSettings` throw an exception because the error was already caught in the action. Using this PR, `saveSettings` will throw an error and the error can be handled. 

I found four places where `saveSettings` was used:


- `handleSubmitCallback` - Step 3 (Configure your product listings).

Before this PR the `saveSettings` could fail and allow to continue to step 4 without saving rates, times or taxes **types**. With this PR if `saveSettings` fails it will not allow to continue as the data has not been saved correctly. 

https://github.com/woocommerce/google-listings-and-ads/blob/0614caa5a6227e5e41b3ae15b3cc7df6442deb57/js/src/setup-mc/setup-stepper/setup-free-listings/index.js#L91-L97

Before this PR:

https://user-images.githubusercontent.com/2488994/179293546-3cc15421-71f5-4dc6-be62-d0a951651e4b.mp4

After this PR

https://user-images.githubusercontent.com/2488994/179300591-b86882ce-7c0f-49b4-8bfe-abe0a7fb08ce.mov

- `handleChangeCallback` - Step 4 (Confirm store requirements). Before this PR, checking the store requirements could fail and allow the user to complete the process. 

https://github.com/woocommerce/google-listings-and-ads/blob/0614caa5a6227e5e41b3ae15b3cc7df6442deb57/js/src/setup-mc/setup-stepper/store-requirements/index.js#L41-L43

Before this PR:

https://user-images.githubusercontent.com/2488994/179303214-78a70c2b-ce41-4ac4-84de-f4759748047f.mp4

After this PR:

https://user-images.githubusercontent.com/2488994/179303657-e02266ab-f34b-4ed3-b019-b5da98c8342a.mp4


- `useAutoSaveSettingsEffect`,    this PR adds the `try` and `catch` in the autosaving function.

https://github.com/woocommerce/google-listings-and-ads/blob/0614caa5a6227e5e41b3ae15b3cc7df6442deb57/js/src/setup-mc/setup-stepper/setup-free-listings/useAutoSaveSettingsEffect.js#L13-L17


- The errors thrown from `allSettled` will be handled by a custom hook that will be created in a future PR, this PR is not changing this behaviour.

https://github.com/woocommerce/google-listings-and-ads/blob/d72e2bd8bca203a8360d74582c082ea14f2b1e88/js/src/edit-free-campaign/index.js#L208-L213

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

### Case 1

1. Checkout `develop` branch or this PR
1. Go to the onboarding process.  
2. Go to step 3 (Configure your product listings). 
2. Provoke an error when saving the settings.  For example, setting an incorrect URL in the following line: 

https://github.com/woocommerce/google-listings-and-ads/blob/d72e2bd8bca203a8360d74582c082ea14f2b1e88/js/src/data/actions.js#L319
3. Choose one of the below options:

![image](https://user-images.githubusercontent.com/2488994/179305207-ac018b2e-66d5-460c-af4e-3fb36ece08ef.png)

4. Click continue.
5. If you are in `develop` branch, it will allow you to go to step 4 without choosing the shipping/rate/tax type. 
6. If you are using this PR, it will not allow you to continue to step 4 because the data has not been saved.

### Case 2

1. Go to the onboarding process.  
2. Go to step 4 (Confirm store requirements).
3. Confirm your phone number.
4. Provoke an error when saving the settings.  For example, setting an incorrect URL in the following line: 

https://github.com/woocommerce/google-listings-and-ads/blob/d72e2bd8bca203a8360d74582c082ea14f2b1e88/js/src/data/actions.js#L319

5. Check all the prelunch options.

![image](https://user-images.githubusercontent.com/2488994/179305958-27b9819a-d507-4c90-bd41-51bb0b28de3d.png)

5. If you are in `develop` branch, it will allow you to complete the process without saving the prelunch options. 
6. If you are using this PR, it will not allow you to complete the process.

### Additional details:
I was thinking to add tests to the "pre-lunch checklist" section but I realized that there is an [ongoing project](pcTzPl-jU-p2) working to update this section so probably those tests will not be valid with the new feature.


<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

> Tweak - Propagate errors for saveSettings

jorgemd24 committed:
Remove try and catch saveSettings

jorgemd24 committed:
Adjust useAutoSaveSettingsEffect to catch the errors

jorgemd24 committed:
Update handleChangeCallback - StoreRequirements to catch the error when savingSettings

jorgemd24 committed:
Remove unused dependency - noop

jorgemd24 committed:
Update SettingsData typedef

jorgemd24 committed:
Add useAutoSaveSettingsEffect tests

jorgemd24 committed:
Add toHaveBeenCalledWith - mockSaveSettings test

puntope commented on js/src/setup-mc/setup-stepper/store-requirements/index.js:
```diff
@@ -36,10 +36,26 @@ export default function StoreRequirements() {
+			setSaveSettings( true );
+		} catch ( error ) {
+			//Create the notice only once
+			if ( isSaveSettings === true ) {
```
💅  Why not just `if ( isSaveSettings ) {`

puntope commented on js/src/setup-mc/setup-stepper/store-requirements/index.js:
```diff
@@ -36,10 +36,26 @@ export default function StoreRequirements() {
 	 * all onboarding accounts are considered unverified phone numbers.
 	 */
 	const [ isPhoneNumberReady, setPhoneNumberReady ] = useState( false );
+	const [ isSaveSettings, setSaveSettings ] = useState( true );
```
💅 Not sure if this is a typo. `isSaveSetting` 

WDYT of const [ settingsSaved, setSettingsSaved ]

puntope commented on js/src/setup-mc/setup-stepper/store-requirements/index.js:
```diff
@@ -120,7 +136,9 @@ export default function StoreRequirements() {
 									loading={ completing }
-									disabled={ ! isReadyToComplete }
+									disabled={
+										! isReadyToComplete || ! isSaveSettings
```
💅  Wondering if we can add `isSaveSettings` condition inside `isReadyToComplete`

puntope commented on js/src/data/actions.js:
```diff
@@ -80,6 +80,21 @@ export function handleFetchError( error, message ) {
+ * @typedef {Object} SettingsData
+ * @property {boolean} [offer_free_shipping] Whether if the merchant offers free shipping.
+ * @property {string} [shipping_rate] Type of the shipping rate. There a three possible values: 'automatic', 'flat' and 'manual'
+ * @property {string} [shipping_time] Type of the shipping time. There are two possible values: 'flat' and 'manual'.
```
💅 One alternative for the enums could be 

`@property {'automatic'|'flat'|'manual'} [shipping_rate] Type of the shipping rate.`

puntope commented on js/src/setup-mc/setup-stepper/store-requirements/index.js:
```diff
@@ -36,10 +36,26 @@ export default function StoreRequirements() {
+		} catch ( error ) {
+			//Create the notice only once
+			if ( isSaveSettings === true ) {
+				createNotice(
+					'error',
+					__(
+						'There was an error trying to save settings. Please try again later.',
+						'google-listings-and-ads'
+					)
+				);
```
💅  This notice is duplicated with the one in `js/src/setup-mc/setup-stepper/setup-free-listings/useAutoSaveSettingsEffect.js` Maybe to consider refactor them in a function like `errorSavingSettingsNotice` or something like that.

jorgemd24 replied:
Thanks for this suggestion, I have tried and I saw the following points: 

1. This needs to be created as react hook. The `createNotice` depends on `useDispatchCoreNotices`.
2. The react hook will need to return a function wrapping the createNotice, so it can be called anywhere in code, otherwise it will complain about it:

`React Hook "useErrorSavingSettingsNotice" is called in function "saveSettingsCallback" that is neither a React function component nor a custom React Hook function`

So I think adding this hook to create the notice will create unnecessary complexity. WDYT

puntope replied:
Good for now... just wondering if will get worse (in terms of code duplication). 

Maybe we can re-define 

'error'  and the message as a constants somewhere. But it can be built in future PRs

jorgemd24 committed:
Add enums to SettingsData typedef

jorgemd24 committed:
Rename isSaveSetting to settingsSaved

jorgemd24 committed:
Add settingsSaved inside isReadyToComplete

jorgemd24 commented:
Thanks, @puntope for reviewing this and your suggestions! I have addressed your comments except for one, that I have some doubts about it. See https://github.com/woocommerce/google-listings-and-ads/pull/1598#discussion_r924319938

puntope reviewed (approved):
✅  LGTM

Thanks for further adjustments. This is ready to go. Good work
